python/ray/dag/compiled_dag_node.py

```diff
@@ -173,6 +174,106 @@ def do_profile_tasks(
         raise
 
 
+@DeveloperAPI
+def do_stream_tasks(
```

Comment:
Can we combine this logic with `do_exec_tasks`?

I think the more codepaths we have here, the more brittle the code becomes.

Reply:
Yeah the first version of the code does that. 

The downside of that is if/else branches in each of the _read()/_compute()/_write()/_exec_operation() methods, which is not quite clean. I get your point that adding a separate code path is brittle, and agree from general principle. Yet I feel reusing `do_exec_tasks` is arguably brittle as well, as people need to think two different kinds of "execution loops" while maintaining a single method. I think the `do_stream_tasks` code path would evolve more after this PR (e.g., supporting overlapping more, supporting shared memory involved operations) and there are likely interface changes which will make it more incompatible with `do_exec_tasks`. My original thinking was that these two code paths will eventually converge, but it is probably better to keep them separate in the beginning.

The other benefit of different execution loops is not to introduce performance regressions for cases where overlapping is disabled. 

Let me know what you think.

Reply:
I get your point but I think in this case we do not want these paths to evolve separately too much, i.e. in one of the following PRs we should support more controls like disabling overlapping by task or specifying the stream to use. For these changes, it will be best to be able to use the same execution loop.

Reply:
OK, I think I can change to that.

Reply:
Done.